aws/lambda/oss_ci_aws_ce_tracking/lambda_function.py

```diff
@@ -0,0 +1,334 @@
+        self.is_dry_run = is_dry_run
+        self.aws_ce_client = boto3.client("ce")
+        self.cc = None  # clickhouse client set in runtime
+        self.granularity = "DAILY"
+
+    def _fetch(self, client: Any, start: str, end: str, granularity="HOURLY"):
+        logger.info(
+            f"Fetching data from aws cost explorer endpoint within time range {start} to {end} ....."
+        )
+        return client.get_cost_and_usage(
+            TimePeriod={
+                "Start": start,
+                "End": end,
+            },
+            Granularity=granularity,
+            Metrics=["UsageQuantity"],
+            GroupBy=[
+                {"Type": "DIMENSION", "Key": "INSTANCE_TYPE"},
+                {"Type": "DIMENSION", "Key": "USAGE_TYPE"},
+            ],
+            Filter={
+                "Dimensions": {
+                    "Key": "SERVICE",
+                    "Values": ["Amazon Elastic Compute Cloud - Compute"],
+                }
+            },
+        )
+
+    def _process_raw_ce_data(self, resp_reults: List[Dict[str, Any]]):
+        filtered = copy.deepcopy(resp_reults)
+        # filter NoInstanceType since it's normally related to other usages
+        for res in filtered:
+            # filted Groups with keys `NoInstanceType``
+            filtered_groups = [
+                group
+                for group in res.get("Groups", [])
+                if "NoInstanceType" not in group.get("Keys", [])
+            ]
+            res["Groups"] = filtered_groups
+
+        # flatten the aws results in to a record that is ready for clickhouse schema conversion
+        res = []
+        for ts_record in filtered:
+            falttened_list = self.flatten_ts_record(ts_record)
+            res.extend(falttened_list)
+        return res
+
+    def flatten_ts_record(self, record: Dict[str, Any]) -> List[Dict[str, Any]]:
+        results = []
+        start = record.get("TimePeriod", {}).get("Start")
+        end = record.get("TimePeriod", {}).get("End")
+        for group in record["Groups"]:
+            new_group = {}
+            new_group["Start"] = start
+            new_group["End"] = end
+            new_group["Keys"] = group.get("Keys", [])
+            usage_quantity = group.get("Metrics", {}).get("UsageQuantity", {})
+            new_group["Amount"] = usage_quantity.get("Amount", 0)
+            new_group["Unit"] = usage_quantity.get("Unit", 0)
+            results.append(new_group)
+        return results
+
+    def to_db_schema(
+        self, record: Dict[str, Any], record_type: str, granularity: str
+    ) -> Optional[Dict[str, Any]]:
+        keys = record.get("Keys", [])
+        startTime = record.get("Start", "").replace("Z", "+00:00")
+        now = datetime.now(timezone.utc).isoformat()
+        if len(keys) < 2:
+            logger.warning(
+                f"Expected two keys from Record, but got {len(record)} keys:{keys}, skipping the record"
```

Comment:
Is there a valid scenario where keys != 2 and you want to continue? Wouldn't it be better to just error out, so we get alerted from failing lambdas and can fix the root cause if this happens?